7531 Assign correct flags to prefetch buffers

benrubson opened the pull request:
Hello,

This PR solves the following issue :
https://www.illumos.org/issues/7531

Some buffers that could be L2ARC eligible are not flagged such, leading to some performance impact.
L2ARC flags are now correctly set, even for prefetch buffers.
User can still control whether or not prefetch buffers are L2-backed using the `l2arc_noprefetch` tunable.

Thank you 👍 

Ben

prakashsurya commented:
@zettabot go

benrubson commented:
@prakashsurya could you please restart the zettabot tests please ?
Many thx 👍

prakashsurya commented:
@zettabot go

ahrens commented on usr/src/uts/common/fs/zfs/dbuf.c:
```diff
@@ -2505,6 +2510,18 @@ dbuf_prefetch(dnode_t *dn, int64_t level, uint64_t blkid, zio_priority_t prio,
 	/*
+	 * Minimum object to check for ARC flags
+	 */
+	db = kmem_cache_alloc(dbuf_cache, KM_SLEEP);
```
I don't think that "faking up" the dbuf is a good idea.  Maybe we can write `_IMPL()` versions of `DBUF_IS_L2CACHEABLE` and `DBUF_IS_L2COMPESSIBLE` that take the underlying data that they need (objset, level, dnode)?

benrubson replied:
Done, we also avoid alloc/free which is good 👍

ahrens reviewed (changes requested):
When we do a demand read of a prefetched block, we should set ARC_FLAG_L2CACHE etc based on the dbuf.  This happens around line 4531 of `arc_read()`.

Is the underlying problem here that the prefetched blocks are being evicted before we do a demand read on them?

Or is the problem that the demand read isn't actually setting the ARC_FLAG_L2CACHEABLE?

Or is the problem something else?

prakashsurya commented:
@zettabot go

benrubson commented:
@prakashsurya could you please relaunch the test please ?
Sounds like test env has been shut down during the build.
Thx 👍

benrubson commented:
@ahrens I'm going to try to find where do these buffers come from.
I think these are prefetched buffers which are evicted before they are demanded (I don't see other possibility), however I will try to confirm.

Anyway, I think that all prefetched buffers should be `ARC_FLAG_L2CACHE` tagged, as for the demanded buffers.
This does not change the default behaviour because if user chooses to skip caching prefetched buffers (`l2arc_noprefetch=1`, default setting), the `ARC_FLAG_L2CACHE` will be removed from prefetched buffers by `arc_read_done()`.
However, this ensures that prefetched buffers get a chance to be L2-backed before being evicted without having been demanded, which is I think what the user expects setting `l2arc_noprefetch` to 0.
In addition, with the `arc_evict_l2_first` and/or `arc_evict_l2_only` flags (#189), these prefetched buffers get a real chance to be L2-backed before their eviction.

Thank you 👍

prakashsurya commented:
@zettabot go

benrubson commented:
@ahrens So I made some tests to see the properties of the impacted buffers, the evicted ones which are not `ARC_FLAG_L2CACHE` flagged (`!HDR_L2CACHE(hdr)`).

To identify buffers seen as already cached, I added the following :
`hdr->b_flags |= ARC_FLAG_CACHED;`
to `arc_read()`, just before the first following line :
`if (*arc_flags & ARC_FLAG_L2CACHE)`
_(strangely ARC_FLAG_CACHED is not used at all)_

OK, so, almost all impacted buffers are prefetched buffers : `HDR_PREFETCH(hdr)`.
And most of them (93%) have been seen as already cached : `hdr->b_flags & ARC_FLAG_CACHED`.

I would have thought this situation as impossible, as `dbuf_read_impl()` sets the `ARC_FLAG_L2CACHE` flag. I then enforced this flag removing the `if (DBUF_IS_L2CACHEABLE(db))` test just before it, but results were almost the same.

OK so I think this means that the impacted buffers were not "hit" by `dbuf_read_impl()`.
But by which function ?
Does all this mean that in some situations demand read isn't correctly setting the `ARC_FLAG_L2CACHE` ?

Thank you 👍

benrubson commented:
@ahrens, dev-team, could we also move forward with this PR please, as it is in some sort related to #7361.
Thank you again 👍

ahrens commented:
I think from your results you are saying that sometimes we do a demand read that didn't pass ARC_FLAG_L2CACHE to arc_read().  What are those calls?

benrubson commented:
Thank you for your answer Matt.
Unfortunately I was not able to clearly find where do these buffers come from.

However what about my previous assumption ?
> I think that all prefetched buffers should be `ARC_FLAG_L2CACHE` tagged, as for the demanded buffers.
This does not change the default behaviour because if user chooses to skip caching prefetched buffers (`l2arc_noprefetch=1`, default setting), the `ARC_FLAG_L2CACHE` will be removed from prefetched buffers by `arc_read_done()`.
However, this ensures that prefetched buffers get a chance to be L2-backed before being evicted without having been demanded, which is I think what the user expects setting `l2arc_noprefetch` to 0.

Don't you agree with it ?
I think that the current behaviour is wrong, `l2arc_noprefetch` is quite useless as prefetched buffers are not `ARC_FLAG_L2CACHE` tagged.
For sure this PR solves the issue.

Thank you 👍

benrubson commented:
@matt, what about my last comment / assumption ?
Thank you very much for your support 👍

ahrens commented on usr/src/uts/common/fs/zfs/sys/dbuf.h:
```diff
@@ -347,6 +347,12 @@ boolean_t dbuf_is_metadata(dmu_buf_impl_t *db);
 	(dbuf_is_metadata(_db) &&					\
 	((_db)->db_objset->os_secondary_cache == ZFS_CACHE_METADATA)))
```
This should use `DBUF_IS_L2CACHEABLE_IMPL()`

benrubson replied:
Something like this @ahrens ?
```
#define	DBUF_IS_L2CACHEABLE(_db) do {						\
	DB_DNODE_ENTER(db);							\
	boolean_t result = DBUF_IS_L2CACHEABLE_IMPL(DB_DNODE(db),db->db_level);	\
	DB_DNODE_EXIT(db);							\
	result;									\
} while (0)
```

ahrens replied:
I guess that doesn't really work (it doesn't evaluate to `result`).  You'd have to convert this to an inline function.

benrubson replied:
Not sure then I really see how to construct such a function without breaking `DBUF_IS_L2CACHEABLE*` working :|

ahrens commented:
> I think from your results you are saying that sometimes we do a demand read that didn't pass ARC_FLAG_L2CACHE to arc_read(). What are those calls?

I think that those calls are buggy, and we should fix them.  You should be able to find them using dtrace, by examining the arguments to `arc_read`, and then printing the stack.

benrubson commented:
DOn't you think that `l2arc_noprefetch=0` is useless because prefetched buffers are not `ARC_FLAG_L2CACHE` flagged ?
If a buffer is prefetched but not hit, it will not be L2ARC eligible, because it will not be flagged accordingly, so `l2arc_noprefetch=0` has no effect on this buffer here :/

benrubson commented:
Hi,

Trying to resume activity on this, I would really like to see this merged.

@ahrens, don't you think that `l2arc_noprefetch=0` is useless because prefetched buffers are not `ARC_FLAG_L2CACHE` flagged ?
If a buffer is prefetched but not hit, it will not be L2ARC eligible, because it will not be flagged accordingly, so `l2arc_noprefetch=0` has no effect on this buffer here :/

Could we then please reconsider this PR ?
Thank you very much 👍 
Ben

benrubson commented:
@prakashsurya could you restart the test please ? Many thx 👍

benrubson commented:
Test passed, thank you for having launched it @prakashsurya 👍 

I'm pretty convinced prefetched buffers should be `ARC_FLAG_L2CACHE` flagged (what this PR currently does), as `l2arc_noprefetch` user tunable will then control whether or not these buffers can be L2 backed.
This gives a nice performance improvement, as shown in the [bug report](https://www.illumos.org/issues/7531).

FreeBSD has _dtrace_, but I never used it, so I've no idea how to easily track the path of these buffers which, according to you @ahrens, should have been flagged by `arc_read`. And what about prefetched buffers which are evicted before being read ?

Thank you very much for your help 👍

ahrens commented:
> l2arc_noprefetch=0 is useless because prefetched buffers are not ARC_FLAG_L2CACHE flagged
If a buffer is prefetched but not hit, it will not be L2ARC eligible, because it will not be flagged accordingly, so l2arc_noprefetch=0 has no effect on this buffer here

I agree; I'll accept this PR based on that.

ahrens reviewed (approved):
It would be nice to clean up DBUF_IS_L2CACHEABLE so that the logic isn't duplicated with DBUF_IS_L2CACHEABLE_IMPL, but I won't make that a requirement.

benrubson commented:
Thank you Matt 👍 
I'm really glad :)

benrubson commented:
Any idea when this tiny patch will be merged ?
Thank you very much 👍

Commit:
Assign correct flags to prefetched buffers

so that they are now L2ARC eligible, leading to some performance improvement (see https://www.illumos.org/issues/7531).

User can still control whether or not these blocks are L2-backed using the l2arc_noprefetch tunable.
